cpp/src/arrow/array/builder_base.cc

```diff
@@ -319,10 +324,30 @@ struct DerefConstIterator {
   pointer operator->() const { return &(**it); }
 };
 
+/// If A and B are equivalent types, a builder of type A can receive
+/// scalar values of type B and a builder of type B can receive
+/// scalar values of type A.
+///
+/// \param a Type A.
+/// \param b Type B.
+bool AreScalarTypesEquivalent(const DataType& a, const DataType& b) {
+  if (a.Equals(b)) {
+    return true;
+  }
+  if ((a.id() == Type::LIST && b.id() == Type::LIST_VIEW) ||
+      (a.id() == Type::LIST_VIEW && b.id() == Type::LIST) ||
+      (a.id() == Type::LARGE_LIST && b.id() == Type::LARGE_LIST_VIEW) ||
+      (a.id() == Type::LARGE_LIST_VIEW && b.id() == Type::LARGE_LIST)) {
+    return checked_cast<const BaseListType&>(a).value_type()->Equals(
+        *checked_cast<const BaseListType&>(b).value_type());
+  }
+  return false;
+}
+
 }  // namespace
 
 Status ArrayBuilder::AppendScalar(const Scalar& scalar, int64_t n_repeats) {
-  if (!scalar.type->Equals(type())) {
+  if (!AreScalarTypesEquivalent(*scalar.type, *type())) {
```

Comment:
Where is this used? I'm not sure I like the idea of introducing implicit conversions. For one thing, what if the builder is `list(field("item", int8(), /*nullable=*/false))` but the appended scalar is `list(int8())`? The implicit conversion would drop the guarantee of no nulls from the built array

Reply:
This simplified work I was doing in compute kernels tremendously. I think that if since we have different physical representation for the same logical types we should start thinking about these conversions. Perhaps not in the ad-hoc way I've done here but with a more robust and general type-checking algorithm.

Reply:
I'm not opposed to the conversion, just to making it a property of the builder. I think if things do break then it'd be possible to recover functionality by extracting the conversion to `template <typename S> Result<S> EquivalentCast(const Scalar&);` (or maybe being more formal and adding `Scalar::View`) and calling that before AppendScalar

Reply:
I added a commit that reverts these changes. All tests passes. I might float this type equivalence idea in the future when I can build a case for it.